Optimize sort followed by limit

jedelbo opened the pull request:
If the limit is much smaller than the total size of the TableView, then it is faster to make a sorted insert into a vector that is kept at the limit size.

## What, How & Why?
<!-- Describe the changes and give some hints to guide your reviewers if possible. -->
<!-- Link to relevant issue this fixes -->

## ☑️ ToDos
* [ ] 📝 Changelog update
* [ ] 🚦 Tests (or not relevant)
* [ ] C-API, if public C++ API changed.

ironage commented:
Once we do https://github.com/realm/realm-core/issues/6933 we should be able to optimize this further by only inserting the _limit_ number of elements into the vector in the first place instead of having an extra buffer like you do here. Given that we expect future optimizations, could you please move the benchmark out of the test and into benchmark-common-tasks so that we can track this case over time. If you could post that benchmark's results compared to current master, I'd also be curious to see what kind of improvement this actually is 😄  
The actual code changes seem fine to me though; this is a sorely needed optimization 👍

tgoyne commented on src/realm/sort_descriptor.cpp:
```diff
@@ -301,7 +301,29 @@ BaseDescriptor::Sorter SortDescriptor::sorter(Table const& table, const IndexPai
+        IndexPairs buffer;
+        buffer.reserve(limit + 1);
+        for (auto& elem : v) {
+            auto it = std::lower_bound(buffer.begin(), buffer.end(), elem, predicate);
```
If `std::ref()` is doing something useful in the `std::sort()` case then it will here too (it makes it so that the predicate is not copied when it's passed to the algorithm).

jedelbo replied:
:+1:

jedelbo commented:
@ironage I am not sure how we can avoid the buffer. We cannot clear the buffer that holds the elements we should sort.
I have added a benchmark. You should be aware the we can get any performance gain we want by carefully selecting the original table view size and limit .In the test I added, the size and limit is 10000 and 100 respectively. This is the result
```
Req runs:  387  SortThenLimit (MemOnly, EncryptionOff):       min   1.26ms (-36.01%)           max   1.33ms (-35.29%)           med   1.28ms (-35.62%)           avg   1.28ms (-35.59%)           stddev    14us (-17.69%)      
```

ironage reviewed (approved):
👍
